ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/routing/RegexRouting.kt

```diff
@@ -0,0 +1,174 @@
+ * @see [Application.routing]
+ */
+@KtorDsl
+@JvmName("putTypedPath")
+public inline fun <reified R : Any> Route.put(
+    path: Regex,
+    crossinline body: suspend PipelineContext<Unit, ApplicationCall>.(R) -> Unit
+): Route = put(path) {
+    body(call.receive())
+}
+
+/**
+ * Builds a route to match `PATCH` requests with the specified regex [path].
+ * @see [Application.routing]
+ */
+@KtorDsl
+public fun Route.patch(path: Regex, body: PipelineInterceptor<Unit, ApplicationCall>): Route {
+    return route(path, HttpMethod.Patch) { handle(body) }
+}
+
+/**
+ * Builds a route to match `PATCH` requests with the specified regex [path] receiving a request body as content of the [R] type.
+ * @see [Application.routing]
+ */
+@KtorDsl
+@JvmName("patchTypedPath")
+public inline fun <reified R : Any> Route.patch(
+    path: Regex,
+    crossinline body: suspend PipelineContext<Unit, ApplicationCall>.(R) -> Unit
+): Route = patch(path) {
+    body(call.receive())
+}
+
+/**
+ * Builds a route to match `DELETE` requests with the specified regex [path].
+ * @see [Application.routing]
+ */
+@KtorDsl
+public fun Route.delete(path: Regex, body: PipelineInterceptor<Unit, ApplicationCall>): Route {
+    return route(path, HttpMethod.Delete) { handle(body) }
+}
+
+/**
+ * Builds a route to match `OPTIONS` requests with the specified regex [path].
+ * @see [Application.routing]
+ */
+@KtorDsl
+public fun Route.options(path: Regex, body: PipelineInterceptor<Unit, ApplicationCall>): Route {
+    return route(path, HttpMethod.Options) { handle(body) }
+}
+
+private fun Route.createRouteFromRegexPath(regex: Regex): Route {
+    return this.createChild(PathSegmentRegexRouteSelector(regex))
+}
+
+public class PathSegmentRegexRouteSelector(private val regex: Regex) : RouteSelector() {
+
+    override fun evaluate(context: RoutingResolveContext, segmentIndex: Int): RouteSelectorEvaluation {
+        val prefix = if (regex.pattern.startsWith('/')) "/" else ""
+        val postfix = if (regex.pattern.endsWith('/')) "/" else ""
+        val pathSegments = context.segments.drop(segmentIndex).joinToString("/", prefix, postfix)
+        return regex.find(pathSegments)?.let { result ->
+            val segmentIncrement = result.value.length.let {
+                if (pathSegments.length == it) {
+                    context.segments.size - segmentIndex
+                } else if (pathSegments[it] == '/') {
+                    val segments = result.value.substring(0, it)
+                    val count = segments.count { c -> c == '/' }
+                    if (prefix == "/") count else count + 1
+                } else {
+                    return RouteSelectorEvaluation.Failed
+                }
+            }
+
+            val groups = result.groups as MatchNamedGroupCollection
+            val parameters = mutableListOf<Pair<String, List<String>>>()
+            GROUP_NAME_MATCHER.findAll(regex.pattern).forEach {
```

Comment:
why do you need to parse regex with regex? can't you get all group names from `result.groups`?

Reply:
I didn't find a way to obtain the names of the named capturing groups in Kotlin. It contains only `value` and `range` of this value in the input. 
I found some ways [here](https://stackoverflow.com/questions/15588903/get-group-names-in-java-regex#:~:text=You%20can%20use%20reflection%20to,(%22namedGroups%22)%3B%20namedGroups.) but they don't seem to be good